src/msgpack/unpacker.d

```diff
@@ -196,7 +196,11 @@ struct Unpacker
         if (0x00 <= header && header <= 0x7f) {
             value = cast(T)header;
         } else if (0xe0 <= header && header <= 0xff) {
-            value = -(cast(T)-header);
+            auto b = cast(T)(cast(ubyte)(-int(header)));
```

Comment:
such a code smell.

Reply:
Is `int()` needed?

Reply:
hmmm... obviously I forgot to paste the article of the compiler change, also don't find it yet, have to search at home. You have no negation on T.sizeof<4 types anymore without providing a certain compilerflag. Therefore everything has to get converted to int for negation. Searched for a way to solve this using binary operators but found none. Since I'm not that experienced in using binary operators, for sure there might be a better way. However, this is the proposed "new" way.

Reply:
integral promotion not done for -header, use '-transition=intpromote' switch or -cast(int)(header)
=> should that be:
`value = -int(cast(T)(-int(header)));`
?

Reply:
related: filed https://issues.dlang.org/show_bug.cgi?id=18307 for a better compiler error message

Reply:
NOTE: pasting the relevant part from https://dlang.org/changelog/2.078.0.html#fix16997
The values computed will be different for some values of the operands:

* All values of - and ~ applied to ubyte, ushort, char, and dchar operands will change.

* The values of -byte(-128) and -short(-32768) will change.

Option 1:

Rewrite op b as op int(b) to use correct, i.e. C-like, behavior.

Option 2:

Rewrite op b as typeof(b)(op int(b)) to preserve the old behavior.

which one is correct behavior? seems like unittest for that would be helpful

Reply:
well msgpack-d has unittests for this exact behavior. and those are passing for this pull request